### ISSUE TITLE
Getting `namedtuple` out of `tfd.JointDistributionCoroutine`

### ISSUE THREAD

**Issue by chrism0dwk** [tensorflow/probability#1517]
Hi Team,

`tfd.JointDistributionCoroutine` represents a draw from its respective joint distribution as a Python `namedtuple`.  Is there any way to access the constructor for that particular `namedtuple` in order to pass an observation to `log_prob`?  Best I've come up with so far is:
```python
def model_fn():
    alpha = yield tfd.Normal(0, 1, name='alpha')
    # ... etc

model = tfd.JointDistributionCoroutine(model_fn)

# Surely there's a better way to construct a sample tuple?
SampleTuple = model.event_shape.__class__
sample = SampleTuple(alpha=0.3, ...)

model.log_prob(sample)
```

Any advances? 

Thanks,

Chris

**Comment by ColCarroll** [comment]
Hey Chris --
You can actually use named kwargs *or* `*args`: 
<img width="465" alt="image" src="https://user-images.githubusercontent.com/2295568/154081869-d57b2052-c13a-4ffb-9483-4c73f55a2557.png">

**Comment by ColCarroll** [comment]
(feel free to reopen/continue discussion -- closing for tidiness)

**Comment by chrism0dwk** [comment]
Thanks both.  Just for context, and returning to the Gibbs question, I'm after splitting a StructTuple  by parameter name within a `GibbsStep.one_step`.  The plan goes something like:
```python
class GibbsStep(tfp.mcmc.TransitionKernel):
    def __init__(self, vars_to_update, model, inner_kernel_fn, kernel_kwargs):
        self.vars_to_update = vars_to_update
        self.model = model
        self.inner_kernel = inner_kernel_fn
        self.kernel_kwargs = kernel_kwargs

    def one_step(self, current_full_state, previous_full_results):

        state_dict = current_full_state._asdict()
        unpinned, pins = _split_dict_into_unpinned_and_pinned(current_full_state, self.vars_to_update)
        previous_results = _get_kernel_results(previous_full_results)

        conditional_model = self.model.experimental_pin(**pins)
        kernel = self.inner_kernel_fn(conditional_model.log_prob, **self.kernel_kwargs)
        packed_unpinned = type(conditional_model.dtype)(**unpinned)  # This is where I need the namedtuple

        next_unpinned, results = kernel.one_step(packed_unpinned, previous_results)
        
        return _repack(next_unpinned, pins, results, model) # Re-pack updated variables into full state tuple
```

Is this totally crazy?

**Comment by brianwa84** [comment]
Looks pretty reasonable to me. I like how you've taken advantage of
experimental_pin for the gibbs step. I suppose not every use-case would do
that e.g. if you can compute a complete conditional and directly sample the
target distribution. But still, many cases like MH within Gibbs might be
addressed using such a recipe.

**Comment by chrism0dwk** [comment]
...feels like the pattern ought to generalise, though.  Either way, you need to close over the pins in order to create a draw from a conditional posterior.  Perhaps it would be possible to swap out a `inner_kernel_fn` for a `make_conditional_distribution_fn` and detect whether a `tfp.mcmc.TransitionKernel` or a `tfd.Distribution` is returned, and then do the right thing.
